File: route-tta-ios/RouteTta/ViewController.m
/*
 * Copyright (c) 2011-2018 HERE Europe B.V.
 * All rights reserved.
 */

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    /* Define waypoints for the route */
    /* START: Holländerstraße, Wedding, 13407 Berlin */
    NMAGeoCoordinates* startPoint = [[NMAGeoCoordinates alloc] initWithLatitude:52.562755700200796 longitude:13.34599438123405];
    
    /* MIDDLE: Lynarstraße 3 */
    NMAGeoCoordinates* middlePoint = [[NMAGeoCoordinates alloc] initWithLatitude:52.54172 longitude:13.36354];
    
    /* END: Agricolastraße 29, 10555 Berlin */
    NMAGeoCoordinates* endPoint = [[NMAGeoCoordinates alloc] initWithLatitude: 52.520720371976495 longitude: 13.332345457747579];
    
    /* Initialize a RoutePlan */
    NSArray* routePlan = [NSArray arrayWithObjects:startPoint, middlePoint, endPoint, nil];
    
    /*
     * Initialize a RouteOption.HERE SDK allow users to define their own parameters for the
     * route calculation,including transport modes,route types and route restrictions etc.Please
     * refer to API doc for full list of APIs
     */
    
    NMARoutingMode* routeMode = [[NMARoutingMode alloc] init];
    /* Other transport modes are also available e.g Pedestrian */
    routeMode.transportMode = NMATransportModeCar;
    /* Disable highway in this route. */
    routeMode.routingOptions = NMARoutingOptionAvoidHighway;
    /* Calculate the shortest route available. */
    routeMode.routingType = NMARoutingTypeFastest;
    /* Calculate 1 route. */
    routeMode.resultLimit = 1;
    
    /* Initialize a CoreRouter */
    _coreRouter = [[NMACoreRouter alloc] init];
    
    // for calculating traffic on the route
    NMADynamicPenalty* penalty = [[NMADynamicPenalty alloc] init];
    penalty.trafficPenaltyMode = NMATrafficPenaltyModeOptimal;
    _coreRouter.dynamicPenalty = penalty;
    
    [_coreRouter calculateRouteWithStops:routePlan routingMode:routeMode completionBlock:
     ^(NMARouteResult * _Nullable routeResult, NMARoutingError error) {
         NMARoute* route = routeResult.routes[0];
         
         // check if route exist
         if (route == nil) {
             return;
         }
         
         self->_calculatedRoute = route;
         
         // add route on the map
         NMAMapRoute* mapRoute = [[NMAMapRoute alloc] initWithRoute:self->_calculatedRoute];
         [self->_mapView addMapObject:mapRoute];
         [self->_mapView setBoundingBox:route.boundingBox withAnimation:NMAMapAnimationNone];
         
         // start downloading traffic for route
         const NMATrafficManager* const manager = [NMATrafficManager sharedTrafficManager];
         [manager requestTrafficOnRoute:route];
         [manager addObserver:self];
         [self setTtaTime];
    }];
}

- (void) viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [[NMATrafficManager sharedTrafficManager] removeObserver:self];
}

- (void) setTtaTime {
    const NSTimeInterval timeIncluded = [_calculatedRoute ttaIncludingTrafficForSubleg:NMARouteSublegWhole].duration;
    const NSTimeInterval timeExcluded = [_calculatedRoute ttaExcludingTrafficForSubleg:NMARouteSublegWhole].duration;
    self->_includedLabel.text = [NSString stringWithFormat:@"ttaIncluded:%.f", timeIncluded];
    self->_excludedLabel.text = [NSString stringWithFormat:@"ttaExcluded:%.f", timeExcluded];
}

- (void)trafficDataDidFinish {
    const NSTimeInterval timeDownloaded = [_calculatedRoute ttaUsingDownloadedTrafficForSubleg:NMARouteSublegWhole].duration;
    self->_downloadedLabel.text = [NSString stringWithFormat:@"ttaDownloaded:%.f", timeDownloaded];
}

@end

File: route-tta-ios-swift/RouteTta/ViewController.swift
/*
 * Copyright (c) 2011-2018 HERE Europe B.V.
 * All rights reserved.
 */

import UIKit
import NMAKit

class ViewController: UIViewController, NMATrafficManagerObserver {

    @IBOutlet weak var ttaExcluded: UILabel?
    @IBOutlet weak var ttaIncluded: UILabel?
    @IBOutlet weak var ttaDownloaded: UILabel?
    @IBOutlet weak var mapView: NMAMapView?
    
    var coreRouter: NMACoreRouter?
    weak var calculatedRoute: NMARoute?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // create route
        let routeContainer = createRoute()
        
        /* Initialize a CoreRouter */
        coreRouter = NMACoreRouter()
        
        // for calculating traffic on the route
        let penalty = NMADynamicPenalty()
        penalty.trafficPenaltyMode = NMATrafficPenaltyMode.optimal
        coreRouter?.dynamicPenalty = penalty
        
        coreRouter?.calculateRoute(withStops: routeContainer.plan, routingMode: routeContainer.mode,
                                   { (result, error) in
            // check error and unwrap route
            guard let route = result?.routes?.first, error == NMARoutingError.none else {
                return
            }
        
            // assign new route
            self.calculatedRoute = route
            // create map object from route
            guard let mapRoute = NMAMapRoute(route) else {
                return
            }
            self.mapView?.add(mapObject: mapRoute)
            self.mapView?.set(boundingBox: route.boundingBox!, animation: NMAMapAnimation.none)
            
            // request for downloading traffic for current route
            let manager = NMATrafficManager.sharedInstance()
            manager.requestTraffic(on: route)
            manager.add(observer: self)
            self.setTtaTime()
        })
    }
    
    func setTtaTime() {
        // get time that includes traffic
        guard let timeIncluded = calculatedRoute?.ttaIncludingTraffic(
            forSubleg: UInt(NMARouteSublegWhole))?.duration else {
            return
        }
        
        // get time without traffic
        guard let timeExcluded = calculatedRoute?.ttaExcludingTraffic(
            forSubleg: UInt(NMARouteSublegWhole))?.duration else {
                return
        }
        
        self.ttaExcluded?.text = "ttaExcluded: \(String(describing: timeExcluded))"
        self.ttaIncluded?.text = "ttaIncluded: \(String(describing: timeIncluded))"
    }
    
    func trafficDataDidFinish() {
        // when downloading of traffic is completed set it to label
        guard let timeDownloaded = calculatedRoute?.ttaUsingDownloadedTraffic(
            forSubleg: UInt(NMARouteSublegWhole))?.duration else {
                return
        }
        
        self.ttaDownloaded?.text = "ttaDownloaded: \(String(describing: timeDownloaded))"
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        NMATrafficManager.sharedInstance().remove(observer: self)
    }
}
